index.ts

```diff
@@ -432,28 +446,236 @@ export class CapabilitiesController extends BaseController<any, any> implements
     });
 
     for (const methodName in newPermissions) {
-      const newPerm = newPermissions[methodName];
-      
-      domain.permissions.push(new Capability({
-        method: newPerm.parentCapability,
-        invoker: domainName,
-        caveats: newPerm.caveats,
-      }));
+      domain.permissions.push(newPermissions[methodName]);
     }
 
     this.setDomain(domainName, domain);
   }
 
+  /**
+   * Validates the given caveats. Returns true if valid, false otherwise.
+   * If the caveats have names, they must be unique.
+   * 
+   * @param {IOcapLdCaveat[]} - The caveats to validate.
+   */
+  validateCaveats (caveats: IOcapLdCaveat[]): boolean {
+
+    const seenNames: { [key: string]: boolean } = {}
+    for (const c of caveats) {
+      if (
+        typeof c !== 'object' || Array.isArray(c) ||
+        !c.type || typeof c.type !== 'string' || 
+        c.name === '' || (
+          c.name && (
+            typeof c.name !== 'string' || seenNames[c.name]
+          )
+        )
+      ) {
+        return false;
+      }
+      if (c.name) {
+        seenNames[c.name] = true;
+      }
+    }
+    return true;
+  }
+
+  /**
+   * Gets all caveats for the permission corresponding to the given domain and
+   * method, or undefined if the permission or its caveats does not exist.
+   * 
+   * @param {string} domainName - The grantee domain.
+   * @param {string} methodName - The name of the method identifying the permission.
+   */
+  getCaveats (
+    domainName: string,
+    methodName: string
+  ): IOcapLdCaveat[] | void {
+
+    const perm = this.getPermission(domainName, methodName);
+    return perm && perm.caveats;
+  }
+
+  /**
+   * Gets the caveat with the given name for the permission corresponding to the
+   * given domain and method, or undefined if the permission or the target
+   * caveat does not exist.
+   * 
+   * @param {string} domainName - The grantee domain.
+   * @param {string} methodName - The name of the method identifying the permission.
+   * @param {string} caveatName - The name of the caveat to retrieve.
+   */
+  getCaveat (
+    domainName: string,
+    methodName: string,
+    caveatName: string
+  ): IOcapLdCaveat | void {
+
+    const perm = this.getPermission(domainName, methodName);
+    return perm
+      ? perm.caveats && perm.caveats.find(c => c.name === caveatName)
+      : undefined;
+  }
+
+  /**
+   * Adds the given caveat to the permission corresponding to the given domain
+   * and method. Throws if the domain or method are unrecognized, or in case of
+   * a caveat name collision.
+   * 
+   * @param {string} domainName - The grantee domain.
+   * @param {string} methodName - The name of the method identifying the permission.
+   * @param {IOcapLdCaveat} caveat - The caveat to add.
+   */
+  addCaveatFor (
+    domainName: string,
+    methodName: string,
+    caveat: IOcapLdCaveat
+  ): void {
+
+    // assert caveat is valid
+    if (!this.validateCaveats([caveat])) {
```

Comment:
See above suggestion for a `validateCaveat` method. That could replace this use of `validateCaveats([caveat])`